libkbfs/rekey_queue.go

```diff
@@ -36,188 +35,92 @@ import (
-
-	wg kbfssync.RepeatedWaitGroup
-
-	lock sync.RWMutex // protects all of the below
-	// pendings tracks TLFs that are in the rekey queue, but doesn't include
-	// those that a worker has already picked up and begun working on.
-	pendings map[tlf.ID]<-chan error
-	// cancel, if non-nil, is for all spawned workers. If nil, no workers should
-	// be running. Calling cancel would cause all workers to stop.
-	cancel context.CancelFunc
+	config  Config
+	log     logger.Logger
+	queue   chan tlf.ID
+	limiter *rate.Limiter
+	cancel  context.CancelFunc
+
+	mu       sync.RWMutex // guards everything below
+	pendings map[tlf.ID]bool
 }
 
 // Test that RekeyQueueStandard fully implements the RekeyQueue interface.
 var _ RekeyQueue = (*RekeyQueueStandard)(nil)
 
-// NewRekeyQueueStandard instantiates a new rekey worker.
-func NewRekeyQueueStandard(config Config) *RekeyQueueStandard {
-	log := config.MakeLogger("RQ")
-	rkq := &RekeyQueueStandard{
+// NewRekeyQueueStandard creates a new rekey queue.
+func NewRekeyQueueStandard(config Config) (rkq *RekeyQueueStandard) {
+	ctx, cancel := context.WithCancel(context.Background())
+	rkq = &RekeyQueueStandard{
 		config:   config,
-		log:      log,
-		queue:    make(chan rekeyQueueEntry, rekeyQueueSize),
-		pendings: make(map[tlf.ID]<-chan error),
+		log:      config.MakeLogger("RQ"),
+		queue:    make(chan tlf.ID, rekeyQueueSize),
+		limiter:  rate.NewLimiter(rekeysPerSecond, numConcurrentRekeys),
+		pendings: make(map[tlf.ID]bool),
+		cancel:   cancel,
 	}
+	rkq.start(ctx)
 	return rkq
 }
 
-// workingOn removes entry from pendings, so that if fbo.Rekey() needs to Enqueue
-// again, it wouldn't be ignored.
-func (rkq *RekeyQueueStandard) workingOn(entry rekeyQueueEntry) {
-	rkq.lock.Lock()
-	defer rkq.lock.Unlock()
-	delete(rkq.pendings, entry.id)
-}
-
-func (rkq *RekeyQueueStandard) doneLocked(entry rekeyQueueEntry, err error) {
-	entry.ch <- err
-	close(entry.ch)
-	delete(rkq.pendings, entry.id)
-	rkq.wg.Done()
-}
-
-func (rkq *RekeyQueueStandard) done(entry rekeyQueueEntry, err error) {
-	rkq.lock.Lock()
-	defer rkq.lock.Unlock()
-	rkq.doneLocked(entry, err)
-}
-
-func (rkq *RekeyQueueStandard) work(ctx context.Context) {
-	for {
-		select {
-		case <-ctx.Done():
-			return
-		case entry := <-rkq.queue:
-			func() {
-				var err error
-				defer rkq.done(entry, err) // deferred in case a panic happens below
-				rkq.workingOn(entry)
-				newCtx := ctxWithRandomIDReplayable(ctx, CtxRekeyIDKey,
-					CtxRekeyOpID, nil)
-				rkq.log.CDebugf(newCtx, "Processing rekey for %s", entry.id)
-				err = rkq.config.KBFSOps().Rekey(newCtx, entry.id)
-			}()
+// start spawns a goroutine that dispatches rekey requests to correct folder
+// branch ops while conforming to the rater limiter.
+func (rkq *RekeyQueueStandard) start(ctx context.Context) {
+	go func() {
+		for id := range rkq.queue {
+			if err := rkq.limiter.Wait(ctx); err != nil {
+				rkq.log.Debug("Waiting on rate limiter for tlf=%v error: %v", id, err)
```

Comment:
Did you intend to continue after the error?  Seems like this should be a panic or a return?